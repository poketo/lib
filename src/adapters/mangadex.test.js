import poketo from '..';
import adapter from './mangadex';

describe('MangadexAdapter', () => {
  describe('supportsUrl', () => {
    it('returns true for urls like mangadex.org', () => {
      expect(adapter.supportsUrl('http://mangadex.org')).toBe(true);
      expect(adapter.supportsUrl('https://www.mangadex.org')).toBe(true);
    });

    it('returns false for urls that are not mangadex.org', () => {
      expect(adapter.supportsUrl('http://maanna.com')).toBe(false);
      expect(adapter.supportsUrl('http://mangadexx.org')).toBe(false);
    });
  });

  describe('parseUrl', () => {
    it('returns the components of a url', () => {
      expect(adapter.parseUrl('https://mangadex.org/title/13127')).toEqual({
        seriesSlug: '13127',
        chapterSlug: null,
      });

      expect(adapter.parseUrl('https://mangadex.org/chapter/47721')).toEqual({
        seriesSlug: null,
        chapterSlug: '47721',
      });
    });

    it('supports legacy MangaDex urls', () => {
      expect(adapter.parseUrl('https://mangadex.org/manga/13127')).toEqual({
        seriesSlug: '13127',
        chapterSlug: null,
      });
    });

    it('throws on unparseable urls', () => {
      expect(() => {
        adapter.parseUrl('https://mangadex.org/about');
      }).toThrow(poketo.InvalidUrlError);
    });
  });
});

describe('MangaDex', () => {
  const server = new AdapterVcrServer(adapter);

  beforeAll(async () => {
    await server.listenAndMock();
  });

  afterAll(() => {
    server.close();
  });

  describe('getSeries', () => {
    it('returns a metadata object', async () => {
      const { chapters, ...metadata } = await poketo.getSeries(
        'mangadex:13127',
      );

      expect(metadata).toMatchSnapshot({
        updatedAt: expect.any(Number),
      });

      const chapterNumbersToTest = ['2', '8'];
      const chaptersToTest = chapters.filter(chapter =>
        chapterNumbersToTest.includes(chapter.chapterNumber),
      );
      expect(chaptersToTest).toMatchSnapshot();
    });

    it('works for series with multiple pages', async () => {
      const longSeries = await poketo.getSeries('mangadex:5411');

      expect(longSeries.chapters.length).toBeGreaterThan(100);
    });

    it('converts html-encoded results', async () => {
      const seriesWithEntities = await poketo.getSeries('mangadex:21117');

      expect(seriesWithEntities.title).toEqual('Robot Ã— Laserbeam');
      expect(seriesWithEntities.description).not.toMatch(/&mdash;/);
    });

    it('only returns the most popular instance of each chapter', async () => {
      const seriesWithMultipleGroups = await poketo.getSeries('mangadex:19729');
      const seriesWithMultipleVolumes = await poketo.getSeries(
        'mangadex:13025',
      );

      const getChapterNumbers = series =>
        series.chapters.map(c => c.chapterNumber);
      const uniq = arr => [...new Set(arr)];

      const a = getChapterNumbers(seriesWithMultipleGroups);
      const b = getChapterNumbers(seriesWithMultipleVolumes);

      expect(a).toEqual(uniq(a));
      expect(b).not.toEqual(uniq(b));
    });
  });

  describe('getChapter', () => {
    it('returns a chapter', async () => {
      const chapter = await poketo.getChapter('mangadex:19915:385894');

      expect(chapter.url).toEqual(`${adapter._getHost()}/chapter/385894`);
      expect(chapter.pages).toHaveLength(1);
      expect(chapter.pages[0]).toEqual(
        expect.objectContaining({
          id: expect.any(String),
          url: expect.stringContaining('http'),
        }),
      );
    });
  });
});
