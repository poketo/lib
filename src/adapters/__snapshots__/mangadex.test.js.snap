// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MangaDex getSeries returns a metadata object 1`] = `
Object {
  "author": "Akitaka",
  "coverImageUrl": "https://mangadex.org/images/manga/13127.large.jpg?1545555212",
  "description": "The Twelve City, a land governed by the leaders of the families who bear in their names the Chinese zodiac. In a city where youkai and humans coexist, the cat youkai who holds a grudge against the Rat Family appears in front of Nekuni Kyouichi, the head of the Rat Family. It's a malice-covered Hateful cat who is aiming for his life. But that youkai is Nekogasaki Natsuho, who's also extremely anxious around other people... Just like the saying \\"Curiosity killed the cat\\" goes, will these feelings also kill the cat?",
  "id": "mangadex:13127",
  "site": Object {
    "id": "mangadex",
    "name": "Mangadex",
  },
  "slug": "13127",
  "status": "ONGOING",
  "supportsReading": true,
  "title": "Uramikoi, Koi, Uramikoi.",
  "updatedAt": Any<Number>,
  "url": "https://mangadex.org/title/13127",
}
`;

exports[`MangaDex getSeries returns a metadata object 2`] = `
Array [
  Object {
    "chapterNumber": "8",
    "createdAt": 1517705219,
    "id": "mangadex:13127:37089",
    "order": 8,
    "slug": "37089",
    "title": "Where the Cat Goes, the Dog will Follow",
    "url": "https://mangadex.org/chapter/37089",
    "volumeNumber": "2",
  },
  Object {
    "chapterNumber": "2",
    "createdAt": 1517704912,
    "id": "mangadex:13127:37060",
    "order": 2,
    "slug": "37060",
    "title": "A Cornered Rat Will Even Bite a Rat",
    "url": "https://mangadex.org/chapter/37060",
    "volumeNumber": "1",
  },
]
`;
